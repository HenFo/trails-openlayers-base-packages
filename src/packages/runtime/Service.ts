export interface ServiceLifecycleHooks {
    /**
     * Destroy the component.
     * Should clean up all resources used by the component.
     */
    destroy?(): void;

    /*
     * Just used so that this interface is not empty.
     * TypeScript would otherwise throw an 'no common properties' error
     * if one tries to implement this interface without providing any
     * lifecycle hooks.
     */
    toString(): string;
}

export type Service<Interface extends {} = {}> = ServiceLifecycleHooks & Interface;

export type ServiceOptions<References extends {} = {}> = {
    /**
     * References to other services.
     * These are injected by the runtime and match the declared references of the service.
     */
    references: References;

    /**
     * Configuration properties.
     *
     * TODO: These are not specified yet
     */
    properties: Record<string, unknown>;
};

export type ServiceConstructor<References extends {} = {}, Interface extends {} = {}> = {
    new (options: ServiceOptions<References>): Service<Interface>;
};
