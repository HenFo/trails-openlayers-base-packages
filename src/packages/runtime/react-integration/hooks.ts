import { useContext, useMemo } from "react";
import { Error } from "@open-pioneer/core";
import { InterfaceName, ServiceType } from "../ServiceRegistry";
import { PackageContext as PackageContext, PackageContextData } from "./PackageContext";
import { ErrorId } from "../errors";

/**
 * Returns a service that implements the given interface.
 * Receives the package name of the importing package as a string.
 *
 * This is an internal hook that is typically called indirectly via the hook
 * provided from `"open-pioneer:react-hooks"`.
 *
 * @private
 */
export function useServiceInternal<Interface extends InterfaceName>(
    packageName: string,
    interfaceName: Interface
): ServiceType<Interface>;
export function useServiceInternal(packageName: string, interfaceName: string): unknown;
export function useServiceInternal(packageName: string, interfaceName: string) {
    const context = useContext(PackageContext);
    const service = useMemo(
        () => checkContext("useService", context).getService(packageName, interfaceName),
        [context, packageName, interfaceName]
    );
    return service;
}

/**
 * Returns the properties for the given package.
 *
 * This is an internal hook that is typically called indirectly via the hook
 * provided from `"open-pioneer:react-hooks"`.
 *
 * @private
 */
export function usePropertiesInternal(packageName: string): Readonly<Record<string, unknown>> {
    const context = useContext(PackageContext);
    return checkContext("useProperties", context).getProperties(packageName);
}

function checkContext(
    hookName: string,
    contextData: PackageContextData | null
): PackageContextData {
    if (!contextData) {
        throw new Error(
            ErrorId.INTERNAL,
            `"Failed to access package context from '${hookName}': react integration was not set up properly.`
        );
    }
    return contextData;
}
