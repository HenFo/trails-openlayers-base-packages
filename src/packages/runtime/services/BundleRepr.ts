import { Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
import { BundleMetadata } from "../Metadata";
import { ServiceRepr } from "./ServiceRepr";

export class BundleRepr {
    static parse(data: BundleMetadata): BundleRepr {
        const name = data.name;
        const services = Object.entries(data.services).map<ServiceRepr>(([name, serviceData]) => {
            if (name !== serviceData.name) {
                throw new Error(
                    ErrorId.INVALID_METADATA,
                    "Invalid metadata: service name mismatch."
                );
            }
            return ServiceRepr.parse(data.name, serviceData);
        });
        return new BundleRepr(name, services);
    }

    readonly name: string;
    readonly services: readonly ServiceRepr[];

    constructor(name: string, services: ServiceRepr[]) {
        this.name = name;
        this.services = services;
    }
}

export function parseBundles(bundles: Record<string, BundleMetadata>) {
    return Object.entries(bundles).map<BundleRepr>(([name, bundleMetadata]) => {
        if (name !== bundleMetadata.name) {
            throw new Error(ErrorId.INVALID_METADATA, "Invalid metadata: bundle name mismatch.");
        }

        return BundleRepr.parse(bundleMetadata);
    });
}
