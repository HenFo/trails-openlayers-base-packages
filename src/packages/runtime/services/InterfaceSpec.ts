export interface InterfaceSpec {
    interfaceName: string;
    qualifier?: string | undefined;
}

export function renderInterfaceSpec({ interfaceName, qualifier }: InterfaceSpec): string {
    return `'${interfaceName}'` + (qualifier ? ` (qualifier: '${qualifier}')` : "");
}
