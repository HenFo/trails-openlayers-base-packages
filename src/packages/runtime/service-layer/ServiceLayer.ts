import { ServiceRepr } from "./ServiceRepr";
import { Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
import {
    ComputedServiceDependencies,
    UIDependency,
    verifyDependencies
} from "./verifyDependencies";
import { PackageRepr } from "./PackageRepr";
import { ReadonlyServiceLookup, ServiceLookupResult, ServicesLookupResult } from "./ServiceLookup";
import {
    InterfaceSpec,
    isAllImplementationsSpec,
    isSingleImplementationSpec,
    ReferenceSpec
} from "./InterfaceSpec";
import { ReferenceMeta } from "../Service";

export type DynamicLookupResult = ServiceLookupResult | UndeclaredDependency;

export interface UndeclaredDependency {
    type: "undeclared";
}

interface DependencyDeclarations {
    all: boolean;
    unqualified: boolean;
    qualifiers: Set<string>;
}

export class ServiceLayer {
    private serviceLookup: ReadonlyServiceLookup;

    // Service id --> Service dependencies
    private serviceDependencies: Map<string, ComputedServiceDependencies>;

    // Package name --> Interface name --> Declarations
    private declaredDependencies;
    private allServices: readonly ServiceRepr[];
    private state: "not-started" | "started" | "destroyed" = "not-started";

    constructor(packages: readonly PackageRepr[]) {
        this.allServices = packages.map((pkg) => pkg.services).flat();
        const { serviceLookup, serviceDependencies } = verifyDependencies({
            services: this.allServices,
            uiDependencies: packages
                .map((pkg) =>
                    pkg.uiReferences.map<UIDependency>((dep) => {
                        return {
                            packageName: pkg.name,
                            ...dep
                        };
                    })
                )
                .flat()
        });
        this.serviceLookup = serviceLookup;
        this.serviceDependencies = serviceDependencies;
        this.declaredDependencies = buildDependencyIndex(packages);
    }

    destroy() {
        this.allServices.forEach((value) => {
            this.destroyService(value);
        });
        this.state = "destroyed";
    }

    start() {
        if (this.state !== "not-started") {
            throw new Error(ErrorId.INTERNAL, "Service layer was already started.");
        }

        this.allServices.forEach((value) => {
            this.createService(value);
        });
        this.state = "started";
    }

    /**
     * Returns a service implementing the given interface.
     * Checks that the given package actually declared a dependency on that interface
     * to enforce coding guidelines.
     *
     * @param packageName the name of the package requesting the import
     * @param spec the interface specifier
     * @param options advanced options to customize lookup
     * @throws if the service layer is not in 'started' state or if no service implements the interface.
     */
    getService(
        packageName: string,
        spec: InterfaceSpec,
        options?: { ignoreDeclarationCheck?: boolean }
    ): ServiceLookupResult | UndeclaredDependency {
        if (this.state !== "started") {
            throw new Error(ErrorId.INTERNAL, "Service layer is not started.");
        }

        if (!options?.ignoreDeclarationCheck && !this.isDeclaredDependency(packageName, spec)) {
            return { type: "undeclared" };
        }

        return this.serviceLookup.lookup(spec);
    }

    /**
     * Returns all services implementing the given interface.
     * Requires that the dependency on all implementations has been declared in the package.
     *
     * @param packageName the name of the package requesting the import
     * @param interfaceName the interface name
     *
     * @throws if the service layer is not in 'started'.
     */
    getServices(
        packageName: string,
        interfaceName: string
    ): ServicesLookupResult | UndeclaredDependency {
        if (this.state !== "started") {
            throw new Error(ErrorId.INTERNAL, "Service layer is not started.");
        }

        if (!this.isDeclaredDependency(packageName, { interfaceName, all: true })) {
            return { type: "undeclared" };
        }

        return this.serviceLookup.lookupAll(interfaceName);
    }

    /**
     * Initializes the given service and its dependencies.
     * Dependencies are initialized before the service that requires them.
     */
    private createService(service: ServiceRepr) {
        if (service.state === "constructed") {
            const instance = service.getInstanceOrThrow();
            service.addRef();
            return instance;
        }
        if (service.state === "constructing") {
            throw new Error(ErrorId.INTERNAL, "Cycle during service construction.");
        }
        if (service.state !== "not-constructed") {
            throw new Error(ErrorId.INTERNAL, "Invalid service state.");
        }

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const references: Record<string, any> = {};
        const referencesMeta: Record<string, ReferenceMeta | ReferenceMeta[]> = {};

        // Sets state to 'constructing' to detect cycles
        service.beforeCreate();

        // Initialize dependencies recursively before creating the current service.
        for (const [referenceName, serviceDeps] of Object.entries(this.getServiceDeps(service))) {
            const [referenceValue, referenceMeta] = this.getReference(serviceDeps);
            references[referenceName] = referenceValue;
            referencesMeta[referenceName] = referenceMeta;
        }

        // Sets state to 'constructed' to finish the state transition, useCount is 1.
        return service.create({ references, referencesMeta, properties: service.properties });
    }

    /**
     * Destroys the given service and its dependencies.
     * The dependencies are destroyed after the service.
     */
    private destroyService(service: ServiceRepr) {
        if (service.state === "destroyed") {
            return;
        }

        // Destroy the service before its dependencies (reverse order
        // compared to construction).
        if (service.removeRef() <= 0) {
            service.destroy();
        }

        for (const serviceDeps of Object.values(this.getServiceDeps(service))) {
            if (Array.isArray(serviceDeps)) {
                for (const dep of serviceDeps) {
                    this.destroyService(dep);
                }
            } else {
                this.destroyService(serviceDeps);
            }
        }
    }

    private isDeclaredDependency(packageName: string, spec: ReferenceSpec) {
        const packageEntry = this.declaredDependencies.get(packageName);
        if (!packageEntry) {
            return false;
        }
        const interfaceEntry = packageEntry.get(spec.interfaceName);
        if (!interfaceEntry) {
            return false;
        }

        if (isSingleImplementationSpec(spec)) {
            if (spec.qualifier == null) {
                return interfaceEntry.unqualified;
            }
            return interfaceEntry.qualifiers.has(spec.qualifier);
        } else {
            return interfaceEntry.all;
        }
    }

    private getReference(ref: ServiceRepr): [unknown, ReferenceMeta];
    private getReference(ref: ServiceRepr[]): [unknown[], ReferenceMeta[]];
    private getReference(
        ref: ServiceRepr | ServiceRepr[]
    ): [unknown, ReferenceMeta] | [unknown[], ReferenceMeta[]];
    private getReference(
        ref: ServiceRepr | ServiceRepr[]
    ): [unknown, ReferenceMeta] | [unknown[], ReferenceMeta[]] {
        if (Array.isArray(ref)) {
            const referenceValues: unknown[] = [];
            const referenceMeta: ReferenceMeta[] = [];
            for (const d of ref) {
                const [value, meta] = this.getReference(d);
                referenceValues.push(value);
                referenceMeta.push(meta);
            }
            return [referenceValues, referenceMeta];
        }

        const referenceValue = this.createService(ref);
        const referenceMeta: ReferenceMeta = {
            serviceId: ref.id
        };
        return [referenceValue, referenceMeta];
    }

    private getServiceDeps(service: ServiceRepr) {
        const dependencies = this.serviceDependencies.get(service.id);
        if (!dependencies) {
            throw new Error(
                ErrorId.INTERNAL,
                `Failed to find precomputed service dependencies for '${service.id}'.`
            );
        }
        return dependencies;
    }
}

function buildDependencyIndex(packages: readonly PackageRepr[]) {
    // Register declared UI dependencies.
    // This is needed as a lookup structure to that dynamic service lookups
    // can be validated at runtime.
    const index = new Map<string, Map<string, DependencyDeclarations>>();
    for (const pkg of packages) {
        const packageName = pkg.name;
        const packageEntry = new Map<string, DependencyDeclarations>();
        for (const uiReference of pkg.uiReferences) {
            let interfaceEntry = packageEntry.get(uiReference.interfaceName);
            if (!interfaceEntry) {
                interfaceEntry = {
                    all: false,
                    unqualified: false,
                    qualifiers: new Set<string>()
                };
                packageEntry.set(uiReference.interfaceName, interfaceEntry);
            }

            if (isAllImplementationsSpec(uiReference)) {
                interfaceEntry.all = true;
            } else if (uiReference.qualifier == null) {
                interfaceEntry.unqualified = true;
            } else {
                interfaceEntry.qualifiers.add(uiReference.qualifier);
            }
        }
        index.set(packageName, packageEntry);
    }
    return index;
}
