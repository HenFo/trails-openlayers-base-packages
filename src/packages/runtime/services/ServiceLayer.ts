import { ServiceRepr } from "./ServiceRepr";
import { Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
import { UIDependency, verifyDependencies } from "./verifyDependencies";
import { PackageRepr } from "./PackageRepr";
import { Service } from "../Service";

export type ServiceLookupResult = FoundService | UnimplementedService | UndeclaredDependency;

export interface FoundService {
    type: "found";
    instance: Service;
}

export interface UndeclaredDependency {
    type: "undeclared";
}

export interface UnimplementedService {
    type: "unimplemented";
}

export class ServiceLayer {
    readonly serviceIndex: ReadonlyMap<string, ServiceRepr>;

    private declaredDependencies = new Map<string, Set<string>>();
    private allServices: readonly ServiceRepr[];
    private state: "not-started" | "started" | "destroyed" = "not-started";

    constructor(packages: readonly PackageRepr[]) {
        this.allServices = packages.map((pkg) => pkg.services).flat();
        this.serviceIndex = verifyDependencies({
            services: this.allServices,
            uiDependencies: packages
                .map((pkg) =>
                    pkg.uiInterfaces.map<UIDependency>((interfaceName) => {
                        return {
                            packageName: pkg.name,
                            interfaceName
                        };
                    })
                )
                .flat()
        });

        for (const pkg of packages) {
            this.declaredDependencies.set(pkg.name, new Set(pkg.uiInterfaces));
        }
    }

    destroy() {
        this.allServices.forEach((value) => {
            this.destroyService(value);
        });
        this.state = "destroyed";
    }

    start() {
        if (this.state !== "not-started") {
            throw new Error(ErrorId.INTERNAL, "Service layer was already started.");
        }

        this.allServices.forEach((value) => {
            this.createService(value);
        });
        this.state = "started";
    }

    /**
     * Returns a service implementing the given interface.
     * Checks that the given package actually declared a dependency on that interface
     * to enforce coding guidelines.
     *
     * @param packageName the name of the package requesting the import
     * @param interfaceName the required interface
     *
     * @throws if the service layer is not in 'started' state or if no service implements the interface.
     */
    getService(packageName: string, interfaceName: string): ServiceLookupResult {
        if (this.state !== "started") {
            throw new Error(ErrorId.INTERNAL, "Service layer is not started.");
        }

        const instance = this.serviceIndex.get(interfaceName)?.instance;
        if (!instance) {
            return { type: "unimplemented" };
        }

        const isDeclared = this.declaredDependencies.get(packageName)?.has(interfaceName) ?? false;
        if (!isDeclared) {
            return { type: "undeclared" };
        }

        return {
            type: "found",
            instance: instance
        };
    }

    /**
     * Initializes the given service and its dependencies.
     * Dependencies are initialized before the service that requires them.
     */
    private createService(service: ServiceRepr) {
        if (service.state === "constructed") {
            const instance = service.getInstanceOrThrow();
            service.addRef();
            return instance;
        }
        if (service.state === "constructing") {
            throw new Error(ErrorId.INTERNAL, "Cycle during service construction.");
        }
        if (service.state !== "not-constructed") {
            throw new Error(ErrorId.INTERNAL, "Invalid service state.");
        }

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const instances: Record<string, any> = {};

        // Sets state to 'constructing' to detect cycles
        service.beforeCreate();

        // Initialize dependencies recursively before creating the current service.
        service.dependencies.forEach((d) => {
            const serviceRef = this.serviceIndex.get(d.interfaceName);
            if (serviceRef) {
                const instance = this.createService(serviceRef);
                instances[d.referenceName] = instance;
            } else {
                throw new Error(ErrorId.INTERNAL, "Service not defined.");
            }
        });

        // Sets state to 'constructed' to finish the state transition, useCount is 1.
        return service.create({ references: instances, properties: service.properties });
    }

    /**
     * Destroys the given service and its dependencies.
     * The dependencies are destroyed after the service.
     */
    private destroyService(service: ServiceRepr) {
        if (service.state === "destroyed") {
            return;
        }

        // Destroy the service before its dependencies (reverse order
        // compared to construction).
        if (service.removeRef() <= 0) {
            service.destroy();
        }

        service.dependencies.forEach((d) => {
            const serviceRef = this.serviceIndex.get(d.interfaceName);
            if (serviceRef) {
                this.destroyService(serviceRef);
            }
        });
    }
}
