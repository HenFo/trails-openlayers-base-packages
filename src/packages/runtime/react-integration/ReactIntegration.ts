import { ComponentType, createElement, StrictMode } from "react";
import { createRoot, Root } from "react-dom/client";
import { Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
import { ServiceLayer } from "../services/ServiceLayer";
import { PackageContext, PackageContextData } from "./PackageContext";
import { PackageRepr } from "../services/PackageRepr";

export interface ReactIntegrationOptions {
    packages: Map<string, PackageRepr>;
    serviceLayer: ServiceLayer;
    rootNode: HTMLDivElement;
}

export class ReactIntegration {
    private packages: Map<string, PackageRepr>;
    private serviceLayer: ServiceLayer;
    private root: Root;
    private packageContext: PackageContextData;

    constructor(options: ReactIntegrationOptions) {
        this.packages = options.packages;
        this.serviceLayer = options.serviceLayer;
        this.root = createRoot(options.rootNode);
        this.packageContext = {
            getService: (packageName, interfaceName) => {
                const result = this.serviceLayer.getService(packageName, interfaceName);
                if (result.type === "unimplemented") {
                    throw new Error(
                        ErrorId.INTERFACE_NOT_FOUND,
                        `The UI of package '${packageName}' requested the unimplemented interface '${interfaceName}'.`
                    );
                }
                if (result.type === "undeclared") {
                    throw new Error(
                        ErrorId.UNDECLARED_DEPENDENCY,
                        `Package '${packageName}' did not declare an UI dependency on interface '${interfaceName}'.` +
                            ` Add the dependency to the package configuration or remove the usage.`
                    );
                }
                return result.instance;
            },
            getProperties: (packageName) => {
                const pkg = this.packages.get(packageName);
                if (!pkg) {
                    throw new Error(
                        ErrorId.INTERNAL,
                        `Package '${packageName}' was not found in application.`
                    );
                }
                return pkg.properties;
            }
        };
    }

    render(component: ComponentType, props: Record<string, unknown>) {
        const element = createElement(component, props);
        const contextWrapper = createElement(
            PackageContext.Provider,
            { value: this.packageContext },
            element
        );
        this.root.render(createElement(StrictMode, undefined, contextWrapper));
    }

    destroy() {
        this.root.unmount();
    }
}
