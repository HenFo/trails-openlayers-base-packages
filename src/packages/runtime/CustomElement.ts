import { ComponentType } from "react";
import { Error, isAbortError, throwAbortError } from "@open-pioneer/core";
import { ErrorId } from "./errors";
import { PackageMetadata } from "./metadata";
import { PackageRepr, createPackages } from "./services/PackageRepr";
import { ServiceLayer } from "./services/ServiceLayer";
import { getErrorChain } from "@open-pioneer/core";
import { ReactIntegration } from "./react-integration/ReactIntegration";

/**
 * Options for the {@link createCustomElement} function.
 */
export interface CustomElementOptions {
    /**
     * Rendered UI component.
     */
    component: ComponentType<Record<string, string>>;

    /**
     * Package metadata.
     * Metadata structures contain information about services
     * that are needed during runtime.
     *
     * Services provided by packages in this attribute will be started
     * as necessary and can be referenced during runtime.
     */
    packages?: Record<string, PackageMetadata>;

    /**
     * Styles for UI component.
     */
    styles?: string;

    /**
     * Application defined properties.
     * These will override default properties of the application's packages.
     */
    properties?: ApplicationProperties;

    /**
     * Function to provide additional application defined properties.
     * Properties returned by this function will override default properties of the application's packages.
     *
     * Compared to {@link properties}, this function receives a context object
     * that allows the developer to provide dynamic properties on a per-application basis.
     *
     * Properties returned by this function take precedence over the ones defined by {@link properties}.
     */
    resolveProperties?(ctx: PropertyContext): Promise<ApplicationProperties>;

    /**
     * Attribute names for component inputs. Changes on this attributes
     * triggers the component rendering.
     */
    attributes?: string[];

    /**
     * Whether the shadow root element is accessible from the outside.
     * @default false
     */
    openShadowRoot?: boolean;
}

/**
 * A context object that is passed to the `resolveProperties` function.
 */
export interface PropertyContext {
    /**
     * Returns an attribute from the application's root node.
     */
    getAttribute(name: string): string | undefined;
}

/**
 * Allows the application to override default properties in all packages.
 */
export interface ApplicationProperties {
    /**
     * Key: the name of the package.
     * Value: A record of configuration properties (key/value pairs).
     *
     * Properties will override default property values in the package.
     */
    [packageName: string]: Record<string, unknown>;
}

/**
 * Creates a new custom element class (web component) that can be registered within a DOM.
 *
 * @example
 * ```ts
 * const CustomElementClazz = createCustomElement({
 *   component: <div>Hello World!</div>,
 *   styles: "div {background-color: red;}",
 * });
 * customElements.define("sample-element", CustomElementClazz);
 * ```
 */
export function createCustomElement(options: CustomElementOptions): CustomElementConstructor {
    class PioneerApplication extends HTMLElement {
        #shadowRoot: ShadowRoot;
        #state: ElementState | undefined;

        static get observedAttributes() {
            return options.attributes ?? [];
        }

        constructor() {
            super();
            this.#shadowRoot = this.attachShadow({
                mode: options.openShadowRoot ? "open" : "closed"
            });
        }

        connectedCallback() {
            if (this.#state) {
                this.#state.destroy();
            }

            this.#state = new ElementState(this, this.#shadowRoot, options);
            this.#state.start();
        }

        disconnectedCallback() {
            this.#state?.destroy();
            this.#state = undefined;
        }

        attributeChangedCallback(name: string, oldValue: string | null, newValue: string | null) {
            this.#state?.onAttributeChanged(name, newValue ?? undefined);
        }
    }
    return PioneerApplication;
}

class ElementState {
    private outerHtmlElement: HTMLElement;
    private shadowRoot: ShadowRoot;
    private options: CustomElementOptions;
    private props: Record<string, string> = {};

    private state = "not-started" as "not-started" | "starting" | "started" | "destroyed";
    private rootNode: HTMLDivElement | undefined;
    private serviceLayer: ServiceLayer | undefined;
    private reactIntegration: ReactIntegration | undefined;

    constructor(
        outerHtmlElement: HTMLElement,
        shadowRoot: ShadowRoot,
        options: CustomElementOptions
    ) {
        this.outerHtmlElement = outerHtmlElement;
        this.shadowRoot = shadowRoot;
        this.options = options;
    }

    start() {
        if (this.state !== "not-started") {
            throw new Error(ErrorId.INTERNAL, `Cannot start element in state '${this.state}'`);
        }

        this.state = "starting";
        this.startImpl().catch((e) => {
            // TODO: Error splash?
            this.destroy();
            if (!isAbortError(e)) {
                logError(e);
            }
        });
    }

    destroy() {
        this.state = "destroyed";
        this.reactIntegration?.destroy();
        this.reactIntegration = undefined;
        this.shadowRoot.replaceChildren();
        this.rootNode = undefined;
        this.serviceLayer?.destroy();
        this.serviceLayer = undefined;
    }

    onAttributeChanged(name: string, value: string | undefined) {
        this.props[name] = value ?? "";
        this.render();
    }

    private async startImpl() {
        const { options, shadowRoot, outerHtmlElement } = this;

        // Resolve custom application properties
        const properties = await gatherProperties(outerHtmlElement, options);
        if (this.state === "destroyed") {
            throwAbortError();
        }

        // Launch the service layer
        const { serviceLayer, packages } = createServiceLayer(options.packages, properties);
        this.serviceLayer = serviceLayer;
        serviceLayer.start();

        // Setup application root node in the shadow dom
        const rootNode = (this.rootNode = document.createElement("div"));
        rootNode.style.height = "100%";
        rootNode.style.width = "100%";
        const style = document.createElement("style");
        style.appendChild(document.createTextNode(options.styles ?? ""));
        shadowRoot.replaceChildren(rootNode, style);

        // Launch react
        this.reactIntegration = new ReactIntegration({
            rootNode: rootNode,
            serviceLayer,
            packages
        });
        this.render();
        this.state = "started";
    }

    private render() {
        this.reactIntegration?.render(this.options.component, this.props);
    }
}

function createServiceLayer(
    packageMetadata: Record<string, PackageMetadata> | undefined,
    properties: ApplicationProperties
) {
    let packages: PackageRepr[];
    try {
        packages = createPackages(packageMetadata ?? {}, properties);
    } catch (e) {
        throw new Error(ErrorId.INVALID_METADATA, "Failed to parse package metadata.", {
            cause: e
        });
    }
    return {
        packages: new Map(packages.map((pkg) => [pkg.name, pkg])),
        serviceLayer: new ServiceLayer(packages)
    };
}

async function gatherProperties(rootNode: HTMLElement, options: CustomElementOptions) {
    let properties: ApplicationProperties[];
    try {
        properties = [
            options.properties ?? {},
            (await options.resolveProperties?.({
                getAttribute(name) {
                    return rootNode.getAttribute(name) ?? undefined;
                }
            })) ?? {}
        ];
    } catch (e) {
        throw new Error(
            ErrorId.PROPERTY_RESOLUTION_FAILED,
            "Failed to resolve application properties.",
            {
                cause: e
            }
        );
    }

    return mergeProperties(properties);
}

function mergeProperties(properties: ApplicationProperties[]): ApplicationProperties {
    const merged: ApplicationProperties = {};
    for (const props of properties) {
        for (const [packageName, packageProperties] of Object.entries(props)) {
            const mergedPackageProps = (merged[packageName] ??= {});
            Object.assign(mergedPackageProps, packageProperties);
        }
    }
    return merged;
}

function logError(e: unknown) {
    if (e instanceof Error) {
        const chain = getErrorChain(e).reverse();
        if (chain.length === 1) {
            console.error(e);
            return;
        }

        let n = 1;
        for (const error of chain) {
            console.error(`#${n}`, error);
            ++n;
        }
    } else {
        console.error("Unexpected error", e);
    }
}
