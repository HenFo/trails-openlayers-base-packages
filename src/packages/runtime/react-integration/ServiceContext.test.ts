import { ReactIntegration } from "./ReactIntegration";
import { ServiceLayer } from "./../services/ServiceLayer";
/**
 * @vitest-environment jsdom
 */
import { createElement } from "react";
import { expect, it } from "vitest";
import { useServiceInternal } from "./hooks";
import { createCustomElement } from "../CustomElement";
import { Service } from "../Service";
import { TestUtils } from "../test/TestUtils";

export interface TestProvider {
    value: string;
}

it("should render component and using service which manipulates dom", async () => {
    function TestComponent() {
        const service = useServiceInternal("test", "test.Provider") as TestProvider;
        return createElement("span", undefined, `Hello ${service.value}`);
    }
    const elem = createCustomElement({
        component: (props) =>
            createElement("div", { id: "wrapper" }, createElement(TestComponent, props)),
        openShadowRoot: true,
        packages: {
            test: {
                name: "test",
                services: {
                    Provider: {
                        name: "Provider",
                        clazz: class Provider implements Service<TestProvider> {
                            value = "TEST";
                        },
                        provides: [
                            {
                                name: "test.Provider"
                            }
                        ]
                    }
                },
                ui: {
                    references: ["test.Provider"]
                }
            }
        }
    });
    customElements.define("test-elem", elem);
    const customElement = await TestUtils.render("test-elem");
    const selectedElem = await TestUtils.waitForSelector("#wrapper", customElement.shadowRoot!);
    expect(selectedElem.querySelector("span")!.innerHTML).toBe("Hello TEST");
});

it("should get error while use undefined service", async () => {
    // TODO: better test for React UI integration
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let resolveCallback: any;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const errorMessage = new Promise<any>((resolve) => {
        resolveCallback = resolve;
    });

    function TestComponent() {
        try {
            const service = useServiceInternal("test", "test.Provider") as TestProvider;
            return createElement("span", undefined, `Hello ${service.value}`);
        } catch (error) {
            resolveCallback(error);
        }
        return null;
    }
    const wrapper = document.createElement("div");
    const serviceLayer = new ServiceLayer([]);
    const reactIntegration = new ReactIntegration({
        rootNode: wrapper,
        serviceLayer
    });
    serviceLayer.start();
    reactIntegration.render(TestComponent, {});
    const error = await errorMessage;
    expect(error.message).toMatchSnapshot();
});
