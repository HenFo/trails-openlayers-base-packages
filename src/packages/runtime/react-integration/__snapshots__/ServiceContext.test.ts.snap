// Vitest Snapshot v1

exports[`should get error while use undefined service 1`] = `"runtime:interface-not-found: The UI of package 'test' requested the unimplemented interface 'test.Provider'."`;
